src/FSharp.Core/prim-types.fs

```diff
@@ -1654,27 +1654,148 @@ namespace Microsoft.FSharp.Core
+                    | _, null -> false
+                    | (:? (obj[])      as arr1), (:? (obj[])      as arr2) -> GenericEqualityObjArray    er iec arr1 arr2
+                    | (:? (byte[])     as arr1), (:? (byte[])     as arr2) -> GenericEqualityByteArray          arr1 arr2
+                    | (:? (int32[])    as arr1), (:? (int32[])    as arr2) -> GenericEqualityInt32Array         arr1 arr2
+                    | (:? (int64[])    as arr1), (:? (int64[])    as arr2) -> GenericEqualityInt64Array         arr1 arr2
+                    | (:? (char[])     as arr1), (:? (char[])     as arr2) -> GenericEqualityCharArray          arr1 arr2
+                    | (:? (float32[])  as arr1), (:? (float32[])  as arr2) -> GenericEqualitySingleArray er     arr1 arr2
+                    | (:? (float[])    as arr1), (:? (float[])    as arr2) -> GenericEqualityDoubleArray er     arr1 arr2
+                    | (:? System.Array as arr1), (:? System.Array as arr2) -> GenericEqualityArbArray    er iec arr1 arr2
+                    | _ -> raise (Exception "invalid logic - expected array")
+
+                let getHashCode (iec, xobj:obj) =
+                  match xobj with 
+                  | null -> 0 
+                  | :? (obj[])      as oa -> GenericHashObjArray iec oa 
+                  | :? (byte[])     as ba -> GenericHashByteArray ba 
+                  | :? (int[])      as ba -> GenericHashInt32Array ba 
+                  | :? (int64[])    as ba -> GenericHashInt64Array ba 
+                  | :? System.Array as a  -> GenericHashArbArray iec a 
+                  | _ -> raise (Exception "invalid logic - expected array")
+
+                { new EqualityComparer<'T>() with
+                    member _.Equals (x, y) = arrayEquals er comparer (box x) (box y)
+                    member _.GetHashCode x = getHashCode (fsEqualityComparerUnlimitedHashingPER, box x) }
+
+            let structuralEqualityComparer<'T> comparer =
+                { new EqualityComparer<'T>() with
+                    member _.Equals (x,y) =
+                        match box x, box y with 
+                        | null, null -> true
+                        | null, _    -> false
+                        | _,    null -> false
+                        | (:? IStructuralEquatable as x1), yobj -> x1.Equals (yobj, comparer)
+                        | _ -> raise (Exception "invalid logic - expected IStructuralEquatable")
+
+                    member _.GetHashCode x =
+                        match box x with 
+                        | null -> 0 
+                        | :? IStructuralEquatable as a -> a.GetHashCode fsEqualityComparerUnlimitedHashingPER
+                        | _ -> raise (Exception "invalid logic - expected IStructuralEquatable") }
+
+            let getEqualityComparer<'T> er =
+                let ty = typeof<'T>
+
+                match er with
+                | _ when canUseDefaultEqualityComparer er ty -> EqualityComparer<'T>.Default
```

Comment:
canUseDefaultEqualityComparer seems like a very big method, are we confident, it won't have a harmful effect on performance

Reply:
it is only called the single time per type... But it will bring some impact to startup times which which is not ideal, but I can't really see any other way. For long running apps this is fine, but for very short lived apps I don't know. I have no idea of what it would mean for AOT compilation? Maybe that should be investigated

Reply:
Some AOT is added now.